autogen/oai/client.py

```diff
@@ -258,16 +273,20 @@ def yes_or_no_filter(context, response):
                             return response
                         continue  # filter is not passed; try the next config
             try:
+                request_ts = self.telemetry.get_current_ts()
                 response = self._completions_create(client, params)
             except APIError as err:
                 error_code = getattr(err, "code", None)
+                self.telemetry.insert(telemetry_id, params, f"error_code:{error_code}, config {i} failed", 0, cleaned_client_config, request_ts)
                 if error_code == "content_filter":
                     # raise the error for content_filter
                     raise
                 logger.debug(f"config {i} failed", exc_info=1)
                 if i == last:
                     raise
             else:
+                self.telemetry.insert(telemetry_id, params, response, 0, cleaned_client_config, request_ts)
+
                 # add cost calculation before caching no matter filter is passed or not
                 response.cost = self.cost(response)
```

Comment:
Thanks for helping with this much needed feature @cheng-tan.
The implementation looks great! 

Two small questions (please feel free to ignore if this is already being addressed, I know the PR is still in draft mode)
- can we log cost since this data might be available 
- do all requests get logged irrespective of caching
- should be have a method to enable/disable logging explicitly .. perhaps something that can be called from conversable agents ... agent.start_logging() etc.

Reply:
Thanks @victordibia,
- cost is available I'll look into adding that to log
- yes, is_cached field indicates if it's reading from cache. 
- yes, will add that shortly

This PR is still WIP but want to get more feedback on the schema, implementation etc. Here are the things I'm planning to look into:
- Enable/disable logging maybe for each agent?
- In agent chat messages adding in the telemetry ID so it can be queried with the telemetry db
- Logging for group chat orchestration
- How does logging work with streaming mode
- Maybe additional things need to be logged with agenteval

Reply:
Thanx to @cheng-tan for this important feature. Two minor comments:

o How hard would it be to add a “verbose” level, maybe something like 0-5 int? Ideal, but not necessary.

o Typo in telemetry.py line 23 - s/b INTEGER.

Reply:
I agree with these recommendations @victordibia 

I also agree that we should add the cost calculations, based on the work @kevin666aa has done here, but perhaps as a separate DB field? I would like to leave the response object as a pure json serialization of whatever the endpoint has returned.

Though not officially supported, we should also test this with some other local llms and hosting solutions (vllm, litellm, lm studio). We can ask our community members to help with this testing. I know this is where we've seen bugs in the past -- and unfortunately when there are uncaught exceptions in this very central part of the code, then basically nothing else can work. To that end, it might be worthwhile being very defensive here, and in telemetry.py so logging issues print errors, but perhaps proceed as if logging was disabled rather than crashing.

Finally, I know we have users who are operating in environments where disks are read-only. Any flags to enable/disable logging should also prevent any database from being initialized -- again, this was an error we previously had with caching, sio it's a bit of a battle scar.